src/hazelcore/UHExp.re

```diff
@@ -308,10 +308,130 @@ let text_operand =
+let associate = (seq: seq): skel => {
+  /**
+   * Shunting-yard opp algorithm according to the specification here
+   * https://en.wikipedia.org/wiki/Shunting-yard_algorithm#The_algorithm_in_detail
+   *
+   */
+  let rec go_seq =
+          (
+            skels: list(skel), /* List of skels to be combined into single output skel. */
+            op_stack: list(Operators_Exp.t), /* Stack of operators. */
+            seq: seq, /* Convert this seq to output skel. */
+            lex_addr: int,
+          ) /* Lexical address of the current operand. */
+          : skel => {
+    switch (seq) {
+    | S(_, affix) =>
+      /**
+       * If the next token is an operand, add a new operand to the output skel.
+       */
+      go_affix(
+        [Skel.Placeholder(lex_addr), ...skels],
+        op_stack,
+        affix, /* Tail of seq without first operand. */
+        lex_addr + 1 /* Increment lexical address of operand in seq. */
+      )
+    };
+  }
+  and go_affix =
+      (
+        skels: list(skel),
+        op_stack: list(Operators_Exp.t),
+        affix: affix,
+        lex_addr: int,
+      )
+      : skel => {
+    switch (affix) {
+    | A(op, seq) =>
+      /**
+       * If the next token is an operator, pop operators from the operator stack,
+       * then push this operator to the operator stack.
+       */
+      let should_mv = op' =>
+        /**
+         * Continue popping operators while the precedence of the top of the operator
+         * stack has greater precedence than the current operator.
+         *
+         * Either the operator on the top of the stack is left associative and
+         * has greater or equal precedence to the current operator, or the
+         * operator on the top of the stack is right associative and has strictly
+         * greater precedence to the current operator.
+         */
+        {
+          switch (Operators_Exp.associativity(op')) {
+          | Left =>
+            Operators_Exp.precedence(op) <= Operators_Exp.precedence(op')
+          | Right =>
+            Operators_Exp.precedence(op) < Operators_Exp.precedence(op')
+          };
+        };
+
+      let (skels', op_stack') = build_ast_while(skels, op_stack, should_mv);
+
+      /* Push this operator to the operator stack. */
+      go_seq(skels', [op, ...op_stack'], seq, lex_addr);
+    | E =>
+      /**
+       * Once the input seq is empty, continuously pop
+       * operators in the stack and build up the output skel.
+       */
+      let (skels', _) = build_ast_while(skels, op_stack, _ => true);
+      switch (skels') {
+      | [final_skel] => final_skel // In this case,
+      | _ => Skel.Placeholder(-1) // This case will never be reached
+      };
+    };
+  }
+  and build_ast_while =
+      (skels: list(skel), op_stack: list(Operators_Exp.t), should_mv)
+      : (list(skel), list(Operators_Exp.t)) => {
+    /* Move operators from the operator stack to the output skel list while... */
+    switch (op_stack) {
```

Comment:
You could compress this `switch` expression and the nested one further below into
```
switch (op_stack, skels) {
| ([op, ...op_stack'], [subskel1, subskel2, ...skels']) when should_mv(op) => ...
| _ => ...
}
```